btree/btree.c

```diff
@@ -3328,104 +3328,112 @@ int calc_shift(int value)
  * btree_create_tick function is the main function used to create btree.
  * It traverses through multiple states to perform its operation.
  *
- * @param smop Represents the state machine operation
- * @return int64_t It returns the next state to be executed.
+ * @param smop     represents the state machine operation
+ * @return int64_t returns the next state to be executed.
  */
-
 int64_t btree_create_tick(struct m0_sm_op *smop)
 {
-	struct m0_btree_op    *bop = M0_AMB(bop, smop, bo_op);
-	struct m0_btree_oimpl *oi = bop->bo_i;
-	struct m0_btree_idata *data = &bop->b_data;
+	struct m0_btree_op    *bop    = M0_AMB(bop, smop, bo_op);
+	struct m0_btree_oimpl *oi     = bop->bo_i;
+	struct m0_btree_idata *data   = &bop->b_data;
 	int                    k_size = data->bt->ksize == -1 ? MAX_KEY_SIZE :
 					data->bt->ksize;
 	int                    v_size = data->bt->vsize == -1 ? MAX_VAL_SIZE :
 					data->bt->vsize;
 
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		/**
+		 * This following check has been added to enforce the
+		 * requirement that nodes have aligned addresses.
+		 * However, in future, this check can be removed if
+		 * such a requirement is invalidated.
+		 */
+		if (!addr_is_aligned(data->addr))
+			return M0_ERR(-EFAULT);
 
-	switch(bop->bo_op.o_sm.sm_state)
-	{
-		case P_INIT:
-			bop->bo_i = m0_alloc(sizeof *bop->bo_i);
-			if (bop->bo_i == NULL)
-				return M0_ERR(-ENOMEM);
-			oi = bop->bo_i;
-			bop->bo_arbor = m0_alloc(sizeof *bop->bo_arbor);
-			if (bop->bo_arbor == NULL)
-				return M0_ERR(-ENOMEM);
-
-			oi->i_nop.no_addr = segaddr_build(data->addr,
-							  calc_shift(
-							  data->num_bytes));
-			return tree_get(&oi->i_nop, &oi->i_nop.no_addr, P_ACT);
+		oi = m0_alloc(sizeof *bop->bo_i);
+		if (oi == NULL)
+			return M0_ERR(-ENOMEM);
```

Comment:
instead of returning M0_ERR(-ENOMEM) you can set bop->bo_op.o_sm.sm_rc = M0_ERR(-ENOMEM), and return P_DONE. as *tick function should return next state which needs to get executed.
Similar changes for tree_destoy, open

Reply:
No, We wont be doing that.
There are changes to the error code return mechanism.
So now all we need to do is return the error.
But I would suggest that you dont make any changes in your PR till these changes are tested thoroughly.

Reply:
okay